### DISCUSSION TITLE
Chat with repo github

### DISCUSSION THREAD

**Discussion by wrapss** [mckaywrigley/chatbot-ui#1159] [Ideas]
I think it might be useful to be able to add a github repo in the files

**Comment by mckaywrigley** [comment]
Totally agree. We're going to be adding "data sources" which are like files for external data sources.

So you'll be able to do retrieval on things like repos, dbs, apis, google drive, notion, etc.

**Reply by jsweetzer-ea** [reply]
Hey Mckay, any updates on this topic? Cheers

**Comment by OgeonX** [comment]
Hey, I just built this. 
<img width="887" alt="image" src="https://github.com/mckaywrigley/chatbot-ui/assets/8281041/857a2c1e-4973-46a7-9623-7b876f2b4c1b">
Is nice and gives me few ideas. 
File upload didnt work, I think should support Azure Storage or any another public cloud storage services providers. 
CGC, AWZ and Azure at least. 
Maybe logins for whatever there is. Social medias, linkedin.  there is many ,


I used cloud supabase and local machine.
I think you could also use Azure web app to host this with node.js or just docker image. 
So it would be online for everyone or maybe not as now it uses my open ai APi key money.
Everybody could have their own chatbot. 
Make .css customizable in the UI. 

Basically this is so easy to build up.
Could also make CICD to Azure Cloud easily. 

<img width="896" alt="image" src="https://github.com/mckaywrigley/chatbot-ui/assets/8281041/da114d50-a562-4907-bf94-beb91e461aaf">

So could have this UI in Azure market place and offer CICD with github with almost click of the button. 
Some are paid as those API calls cost. 

Would be nice to see this online and people could create their owns and develop.

Sorry if this is in wrong place.

**Comment by kinthaiofficial** [comment]
Chat with a GitHub repo is doable with chatbot-ui, but requires some setup around the indexing step. The general approach:

**1. Fetch and index the repo content**
Clone the repo, walk the file tree, chunk files into ~500 token segments (respecting function/class boundaries where possible), embed each chunk, and store in your vector DB (Supabase pgvector works well with chatbot-ui).

```bash
git clone https://github.com/owner/repo /tmp/target-repo
# Then run your indexing script to embed and store
```

**2. Add a "repo context" feature to the chat**
When the user asks a question, retrieve the top-K most relevant chunks from the vector store, inject them into the system prompt as context, then let the LLM answer.

**3. Keep the index fresh**
For active repos, you want incremental re-indexing on push. Set up a GitHub webhook → re-embed changed files → upsert into vector store.

**Practical limits:**
- Large monorepos won't fit in context even with retrieval — limit to the most relevant files per query
- Code understanding degrades for cross-file reasoning (e.g., "where is this function called?") — consider adding a code graph layer for reference tracking
- Binary files, generated code, and vendored dependencies should be excluded from indexing

**Alternatives worth evaluating:**
- SourceGraph Cody has this natively with better code-aware chunking
- Cursor handles this in the IDE with more sophisticated symbol-level retrieval

For a chat-with-repo proof of concept though, chatbot-ui + pgvector + a simple indexing script gets you 80% of the way there without much effort.
